homeassistant/components/fireservicerota/__init__.py

```diff
@@ -0,0 +1,182 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN, CONF_URL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.util import Throttle
+
+from .const import DOMAIN, NOTIFICATION_AUTH_ID, NOTIFICATION_AUTH_TITLE, WSS_BWRURL
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORTED_PLATFORMS = {SENSOR_DOMAIN}
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up the FireServiceRota component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up FireServiceRota from a config entry."""
+    coordinator = FSRDataUpdateCoordinator(hass, entry)
+    await coordinator.async_update()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for platform in SUPPORTED_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload FireServiceRota config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        del hass.data[DOMAIN][entry.entry_id]
+
+    return unload_ok
+
+
+class FSRDataUpdateCoordinator:
+    """Getting the latest data from fireservicerota."""
+
+    def __init__(self, hass, entry):
+        """Initialize the data object."""
+        self._hass = hass
+        self._entry = entry
+
+        self._url = entry.data[CONF_URL]
+        self._tokens = entry.data[CONF_TOKEN]
+
+        self.availability_data = None
+        self.incident_data = None
+        self.incident_id = None
+        self.response_data = None
+
+        self.fsr_avail = FireServiceRota(
+            base_url=f"https://{self._url}", token_info=self._tokens
+        )
+
+        self.fsr_incidents = FireServiceRotaIncidents(on_incident=self.on_incident)
+
+        self.start_listener()
+
+    def construct_url(self) -> str:
+        """Return url with latest config values."""
+        return WSS_BWRURL.format(
+            self._entry.data[CONF_URL], self._entry.data[CONF_TOKEN]["access_token"]
+        )
+
+    def on_incident(self, data) -> None:
+        """Update the current data."""
+        _LOGGER.debug("Got data from websocket listener: %s", data)
+        self.incident_data = data
+
+        async_dispatcher_send(self._hass, f"{DOMAIN}_{self._entry.entry_id}_update")
+
+    def start_listener(self) -> None:
+        """Start the websocket listener."""
+        _LOGGER.debug("Starting incidents listener")
+        self.fsr_incidents.start(self.construct_url())
```

Comment:
Is this doing I/O ?

Reply:
Also, why is the WebSocket url being passed in? Seems odd.

Reply:
It's not doing I/O, it starts a web-socket client as listener in a separate thread.
The auth token is part of the web-socket URL, so it can change during operation (after refreshing tokens)
It restart the web-socket thread with the new URL/token, this only happens every 14 days.

Reply:
This is also the reason I couldn't get the builtin home assistant oauth mechanism to work.

Reply:
Can you elaborate on that? I don't fully understand how the websocket url impacts OAuth

Or maybe easier to reach out to me on Discord.

Reply:
Correction, I couldn't use the HA oauth2 mechanism -or at least find out how yet- because I need to use LegacyApplicationClient type requests like below so with one endpoint, so no callbacks etc.

And I need to know when tokens are refreshed and have them available in code (for restarting the websocket client)
But that last part can be done as far as I could see.

The code I use now works fine, and I don't need to store user credentials this way.
But I can move to HA code later, if I have more information, how it can be done.

```    '''Get tokens'''
        oauth_client = oauthlib.oauth2.LegacyApplicationClient(client_id=None)
        request_body = oauth_client.prepare_request_body(
            username=self._username, password=self._password
        )
        response = self._request(
            "POST",
            endpoint=FSR_ENDPOINT_TOKEN,
            log_msg_action="request tokens",
            params=str.encode(request_body),
            auth_request=True,
        )

        try:
            response = requests.request(
                method,
                url,
                headers=headers,
                params=params,
                json=body,
                timeout=FSR_DEFAULT_TIMEOUT,
            )
            response.raise_for_status()
            return response.json()

         '''Requests with tokens'''
         if not auth_request:
             url = f"{self._base_url}/api/v2/{endpoint}"
             headers = {
                "Content-Type": "application/json;charset=UTF-8",
                "Authorization": f"Bearer {self._token_info['access_token']}",
             }
         try:
            response = requests.request(
                method,
                url,
                headers=headers,
                params=params,
                json=body,
                timeout=FSR_DEFAULT_TIMEOUT,
             )
         response.raise_for_status()
         return response.json()
```
Will have a eye out on discord.

Reply:
Ah yes, we do not support that type of application client.